normcap: init at 0.4.4

cafkafk opened the pull request:
## Description of changes

Closes: #166228

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cafkafk committed:
normcap: init at 0.4.4

Signed-off-by: Christina Sørensen <christina@cafkafk.com>

TheBrainScrambler commented on pkgs/applications/graphics/normcap/default.nix:
```diff
@@ -0,0 +1,43 @@
+  ];
+
+  postPatch = ''
+    sed -i 's|PySide6-Essentials = "6.5.1"||' pyproject.toml
```
Use `substituteInPlace` instead.
https://nixos.org/manual/nixpkgs/stable/#fun-substitute

cafkafk replied:
I tried in the last push, can you check that it's correct?

TheBrainScrambler replied:
Yes, perfect !

TheBrainScrambler reviewed (changes requested):
Not sure if anything else is correct, but I saw that low-hanging fruit

TheBrainScrambler commented:
Also why is this a draft ? Does the package fail to build ?

cafkafk commented:
> Also why is this a draft ? Does the package fail to build ?

It seems to build on ofborg, but I couldn't test locally because cache.nixos.org doesn't seem to have a dependency I needed and building it took an unreasonable amount of time, so I was waiting for it to be cached.

If anyone else can test this locally and report back if it's working I'd mark it ready for review.

TheBrainScrambler commented:
Ah yes shiboken6, I saw that an update for it merged recently. I'm building it right now.

TheBrainScrambler commented on pkgs/applications/graphics/normcap/default.nix:
```diff
@@ -0,0 +1,43 @@
+  meta = with lib; {
+    description = "OCR powered screen-capture tool to capture information instead of images";
+    homepage = "https://dynobo.github.io/normcap/";
+    license = licenses.gpl3;
```
`gpl3` is deprecated. According to https://github.com/dynobo/normcap/blob/main/LICENSE this should be `gpl3Plus`

TheBrainScrambler commented on pkgs/applications/graphics/normcap/default.nix:
```diff
@@ -0,0 +1,43 @@
+    description = "OCR powered screen-capture tool to capture information instead of images";
+    homepage = "https://dynobo.github.io/normcap/";
+    license = licenses.gpl3;
+    maintainers = with maintainers; [cafkafk];
```
`[cafkafk]` -> `[ cafkafk ]`

TheBrainScrambler reviewed (changes requested):
As usual I got a bit fast on my review

TheBrainScrambler commented:
Well then there was pyside6, and I confirm that this one takes too long ... So I will do like you did, not building it and waiting for it to be in cache.

TheBrainScrambler commented:
Result of `nixpkgs-review pr 253578` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>normcap</li>
    <li>normcap.dist</li>
  </ul>
</details>

TheBrainScrambler commented:
pyside6 is finally cached so it's possible to build this now. For some reasons i also get a `normcap.dist` package for free ?

When running normcap though it fails with this error:

```
15:20:57 - CRITICAL - normcap:161 - Uncaught exception! Quitting NormCap!
15:20:57 - CRITICAL - normcap:187 - Uncaught exception! Quitting NormCap! (debug output limited)
Traceback (most recent call last):
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/bin/.normcap-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/app.py", line 57, in main
    tray = SystemTray(app, vars(args))
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/gui/tray.py", line 60, in __init__
    logger.debug("System info:\n%s", system_info.to_dict())
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/gui/system_info.py", line 165, in to_dict
    "tesseract_path": get_tesseract_path(),
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/gui/system_info.py", line 81, in get_tesseract_path
    raise RuntimeError(
RuntimeError: No Tesseract binary found! Tesseract has to be installed and added to PATH environment variable.
15:20:57 - CRITICAL - normcap:192 - Please open an issue with the output above on https://github.com/dynobo/normcap/issues
```

Obviously it's missing the `tesseract` dependency, so you will have to add that.
To avoid more problems like that have a look at the AUR package at https://aur.archlinux.org/packages/normcap and its dependencies, this might help.

cafkafk commented:
I added `tesseract4`, it seems to work now (added `leptonica` as well, but it seemed to work without it). Also added `wl-clipboard` for Wayland users, but I can't test that myself. This does work now, but on my very janky XMonad it can be a bit weird, and I'll likely have to add some fixes locally to make it neat.

Ideally someone would test this on:
- [ ] Wayland, does the clipboard work?
- [ ] KDE, Gnome, or mainstream DE.

> For some reasons i also get a normcap.dist package for free ?

I've got no idea why tbh, hopefully some reviewer can look into this.

TheBrainScrambler commented:
I've tried again `nixpkgs-review` with it and it still fails with this error saying tesseract wasn't found. Looking at your nix file it's because you put tesseract4 in `nativeBuildInputs`. Packages in `nativeBuildInputs` are only available during build time. You have to move it (and leptonica too I guess) to `propagatedBuildInputs`.

I suspect it worked for you because you used a nix-shell ? Next time try it with nix-build to be sure

cafkafk commented:
> I've tried again `nixpkgs-review` with it and it still fails with this error saying tesseract wasn't found. Looking at your nix file it's because you put tesseract4 in `nativeBuildInputs`. Packages in `nativeBuildInputs` are only available during build time. You have to move it (and leptonica too I guess) to `propagatedBuildInputs`.
> 
> I suspect it worked for you because you used a nix-shell ? Next time try it with nix-build to be sure. You can also use nixpkgs-review.

I updated it. Ran `nix build ./#normcap` and ran the binary `./result/bin/normcap`, and it worked. I'm not using a nix-shell thou, so I do wonder why it worked, before I tested via `nix profile install ./#normcap`.

TheBrainScrambler commented:
Result of `nixpkgs-review pr 253578` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>normcap</li>
    <li>normcap.dist</li>
  </ul>
</details>

TheBrainScrambler commented:
Strange, maybe you have tesseract installed and it leaked in ? I don't know.
In any case it works for me too now, so it should be good.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1104

pbsds commented on pkgs/applications/graphics/normcap/default.nix:
```diff
@@ -0,0 +1,52 @@
+  };
+
+  buildInputs = [
+    wl-clipboard
```
Does this include the x11 fallback tools?

cafkafk replied:
¯\\_(ツ)_/¯ I have no idea tbh

tennox commented:
```
$ normcap
16:53:05 - ERROR   - normcap:211 - [QT] qtfatalmsg - this application failed to start because no qt platform plugin could be initialized. reinstalling the application may fix this problem.

available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, xcb.

fish: Job 1, 'normcap' terminated by signal SIGABRT (Abort)
```

Am I doing something wrong?
Or is this an issue?

TheBrainScrambler commented:
I have no idea, for me it works ...
Can't be a regression though since there hasn't been any update for this
package.

cafkafk commented:
> ```
> $ normcap
> 16:53:05 - ERROR   - normcap:211 - [QT] qtfatalmsg - this application failed to start because no qt platform plugin could be initialized. reinstalling the application may fix this problem.
> 
> available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, xcb.
> 
> fish: Job 1, 'normcap' terminated by signal SIGABRT (Abort)
> ```
> 
> Am I doing something wrong? Or is this a regression?

Might wanna create a issue, I can't reproduce this.

yorickvP commented:
wayland support, I bet, try `QT_QPA_PLATFORM=xcb normcap`

tennox commented:
> wayland support, I bet, try `QT_QPA_PLATFORM=xcb normcap`

Thanks, @yorickvP - that fixes it.
Is there a way to bake this into the nixpkgs package?
Found many different ways this env var is set, and tried to find those that are also a python package, but the python ones only set it to e.g. `offscreen` in the check/install phase:
- [vorta](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/backup/vorta/default.nix#L71) (setting in preCheck)
Other apps use e.g.:
- [`qtWrapperArgs`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/video/plex-media-player/default.nix#L42)
- [`wrapProgram`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/viber/default.nix#L82)

I guess the latter might be most promising, but I'm fishing in unknown waters here :thinking:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-properly-set-qt-qpa-platform-xcb-for-an-app/35747/1
